pkg/internal/telemetry/submitter.go

```diff
@@ -0,0 +1,156 @@
+package telemetry
+
+import (
+	"bytes"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"google.golang.org/protobuf/proto"
+
+	"github.com/prefab-cloud/prefab-cloud-go/pkg/internal"
+	"github.com/prefab-cloud/prefab-cloud-go/pkg/internal/contexts"
+	"github.com/prefab-cloud/prefab-cloud-go/pkg/internal/options"
+	prefabProto "github.com/prefab-cloud/prefab-cloud-go/proto"
+)
+
+var NowProvider = time.Now().UnixMilli
+
+type Submitter struct {
+	aggregators                 []Aggregator
+	contextAggregator           Aggregator
+	evaluationSummaryAggregator *EvaluationSummaryAggregator
+	instanceHash                string
+	host                        string
+	apiKey                      string
+	mutex                       *sync.Mutex
+}
+
+type Payload = prefabProto.TelemetryEvents
+
+func NewTelemetrySubmitter(options options.Options) *Submitter {
+	aggregators := []Aggregator{}
+
+	contextAggregator := NewContextAggregator(options)
+
+	if contextAggregator != nil {
+		aggregators = append(aggregators, contextAggregator)
+	}
+
+	var evaluationSummaryAggregator *EvaluationSummaryAggregator
+	if options.CollectEvaluationSummaries {
+		evaluationSummaryAggregator = NewEvaluationSummaryAggregator()
+		aggregators = append(aggregators, evaluationSummaryAggregator)
+	}
+
+	return &Submitter{
+		aggregators:                 aggregators,
+		host:                        options.TelemetryHost,
+		apiKey:                      options.APIKey,
+		contextAggregator:           contextAggregator,
+		evaluationSummaryAggregator: evaluationSummaryAggregator,
+		mutex:                       &sync.Mutex{},
+		instanceHash:                options.InstanceHash,
+	}
+}
+
+func (ts *Submitter) StartPeriodicSubmission(interval time.Duration) {
+	ticker := time.NewTicker(interval)
+
+	go func() {
+		for range ticker.C {
+			if len(ts.aggregators) > 0 {
+				ts.Submit()
+			}
+		}
+	}()
+}
+
+func (ts *Submitter) RecordEvaluation(data internal.ConfigMatch) {
+	if ts.evaluationSummaryAggregator == nil || !data.IsMatch {
+		return
+	}
+
+	// We don't track log level evaluations
+	if _, ok := data.Match.GetType().(*prefabProto.ConfigValue_LogLevel); ok {
+		return
+	}
+
+	ts.evaluationSummaryAggregator.Record(data)
+}
+
+func (ts *Submitter) RecordContext(data *contexts.ContextSet) {
+	if ts.contextAggregator == nil {
+		return
+	}
+
+	ts.contextAggregator.Record(data)
+}
+
+func (ts *Submitter) Submit() error {
+	ts.mutex.Lock()
+	defer ts.mutex.Unlock()
+
+	if len(ts.aggregators) == 0 {
+		return errors.New("no aggregators to submit")
+	}
+
+	payload := Payload{
+		InstanceHash: ts.instanceHash,
+	}
+
+	for _, aggregator := range ts.aggregators {
+		aggregator.Lock()
+
+		data := aggregator.GetData()
```

Comment:
Thinking out loud -- this code locks the aggregator, meaning Get calls are blocked while the the payloads are being built. This code doesn't do any i/o so it should be fast but i do find myself wondering if it would be better to put queue (buffered channels) behind the RecordContext/RecordEvaluation etc calls so that the internal bookkeeping of each aggregator plus this occasional submitter lock don't hold up the caller in any way?

Reply:
Great catch. Handled in 75c777c